File: deploy_to_allinkl.py
#!/usr/bin/env python3

import ftplib
import os
from pathlib import Path
import time

print("===============================================")
print("Metropol Portal - Automatisches Deployment")
print("===============================================\n")

# FTP-Konfiguration
FTP_CONFIG = {
    'host': 'w019e3c7.kasserver.com',
    'user': 'w019e3c7',
    'password': '2Brands2025!',
    'root': '/w019e3c7/firmenpro.de'
}

# MySQL-Konfiguration
DB_CONFIG = {
    'host': 'localhost',
    'database': 'd0446399',
    'user': 'd0446399',
    'password': '2Brands2025!'
}

# Projekt-Pfade
LOCAL_ROOT = Path(__file__).parent
REMOTE_ROOT = FTP_CONFIG['root']

# Verzeichnisse zum Erstellen
DIRECTORIES = ['cache', 'logs', 'uploads', 'temp', 'backups']

# Verzeichnisse zum Hochladen
UPLOAD_DIRS = ['src', 'public', 'config', 'database', 'templates', 'lang', 'routes', 'scripts']

# Dateien die NICHT hochgeladen werden
EXCLUDE_PATTERNS = {
    '.git', '.gitignore', '.github', '.DS_Store', 'node_modules',
    'tests', 'docs', '*.log', '*.md', 'deploy_to_allinkl.py',
    'deploy-to-allinkl.php', 'tsconfig.json', 'package.json',
    'package-lock.json', 'composer.lock', 'phpunit.xml',
    'phpcs.xml', 'phpstan.neon', '__pycache__', '*.pyc'
}

def should_exclude(filename):
    """Prüft ob Datei ausgeschlossen werden soll"""
    for pattern in EXCLUDE_PATTERNS:
        if pattern.startswith('*.'):
            if filename.endswith(pattern[1:]):
                return True
        elif filename == pattern:
            return True
    return False

def upload_file(ftp, local_path, remote_path):
    """Einzelne Datei hochladen"""
    try:
        with open(local_path, 'rb') as file:
            ftp.storbinary(f'STOR {remote_path}', file)
        print(f"✓ Uploaded: {remote_path}")
    except Exception as e:
        print(f"✗ Failed: {remote_path} - {str(e)}")

def create_directory(ftp, path):
    """Verzeichnis erstellen wenn nicht vorhanden"""
    try:
        ftp.mkd(path)
        print(f"✓ Created directory: {path}")
    except ftplib.error_perm as e:
        if "File exists" not in str(e):
            # Versuche parent directory zu erstellen
            parent = os.path.dirname(path)
            if parent and parent != '/':
                create_directory(ftp, parent)
                try:
                    ftp.mkd(path)
                    print(f"✓ Created directory: {path}")
                except:
                    pass

def upload_directory(ftp, local_dir, remote_dir):
    """Verzeichnis rekursiv hochladen"""
    create_directory(ftp, remote_dir)
    
    for item in os.listdir(local_dir):
        if should_exclude(item):
            continue
            
        local_path = os.path.join(local_dir, item)
        remote_path = f"{remote_dir}/{item}"
        
        if os.path.isdir(local_path):
            upload_directory(ftp, local_path, remote_path)
        else:
            upload_file(ftp, local_path, remote_path)

def main():
    # Phase 1: FTP-Verbindung und Upload
    print("Phase 1: FTP-Upload")
    print("-" * 40)
    
    try:
        # FTP-Verbindung
        ftp = ftplib.FTP(FTP_CONFIG['host'])
        ftp.login(FTP_CONFIG['user'], FTP_CONFIG['password'])
        print(f"✓ Connected to {FTP_CONFIG['host']}")
        
        # Basis-Verzeichnisse erstellen
        print("\nCreating base directories...")
        # Erst sicherstellen dass Root existiert
        create_directory(ftp, REMOTE_ROOT)
        
        # Dann Unterverzeichnisse
        for directory in DIRECTORIES:
            create_directory(ftp, f"{REMOTE_ROOT}/{directory}")
        
        # Projekt-Dateien hochladen
        print("\nUploading project files...")
        for directory in UPLOAD_DIRS:
            local_dir = LOCAL_ROOT / directory
            if local_dir.exists():
                print(f"\nUploading /{directory}...")
                upload_directory(ftp, str(local_dir), f"{REMOTE_ROOT}/{directory}")
        
        # composer.json hochladen
        composer_file = LOCAL_ROOT / 'composer.json'
        if composer_file.exists():
            upload_file(ftp, str(composer_file), f"{REMOTE_ROOT}/composer.json")
        
        # Berechtigungen setzen (soweit möglich via FTP)
        try:
            for directory in DIRECTORIES:
                ftp.voidcmd(f"SITE CHMOD 755 {REMOTE_ROOT}/{directory}")
        except:
            print("Note: Could not set permissions via FTP")
        
        ftp.quit()
        print("\n✓ FTP upload completed!")
        
    except Exception as e:
        print(f"\n✗ FTP Error: {str(e)}")
        return
    
    # Phase 2: Datenbank-Setup
    print("\n\nPhase 2: Database Setup")
    print("-" * 40)
    
    sql_file = LOCAL_ROOT / 'database' / 'migrations' / 'create_all_tables.sql'
    if not sql_file.exists():
        print("✗ SQL file not found!")
        return
    
    try:
        # MySQL-Verbindung über SSH-Tunnel wäre nötig
        # Da wir keinen direkten Zugriff haben, erstellen wir ein PHP-Setup-Skript
        setup_script = """<?php
// Temporäres Setup-Skript - NACH AUSFÜHRUNG LÖSCHEN!
error_reporting(E_ALL);
ini_set('display_errors', 1);

$config = [
    'host' => 'localhost',
    'name' => 'd0446399',
    'user' => 'd0446399',
    'password' => '2Brands2025!'
];

try {
    $pdo = new PDO(
        "mysql:host={$config['host']};dbname={$config['name']};charset=utf8mb4",
        $config['user'],
        $config['password'],
        [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION]
    );
    
    // SQL-Datei ausführen
    $sql = file_get_contents(__DIR__ . '/database/migrations/create_all_tables.sql');
    $statements = array_filter(array_map('trim', explode(';', $sql)));
    
    foreach ($statements as $statement) {
        if (!empty($statement)) {
            $pdo->exec($statement);
        }
    }
    
    // Admin-Passwort setzen
    $adminPassword = password_hash('Admin2025!', PASSWORD_BCRYPT, ['cost' => 12]);
    $pdo->exec("UPDATE users SET password = '$adminPassword' WHERE email = 'admin@firmenpro.de'");
    
    echo "✓ Database setup completed!<br>";
    echo "✓ Admin user: admin@firmenpro.de<br>";
    echo "✓ Password: Admin2025!<br>";
    echo "<br><strong>WICHTIG: Löschen Sie diese Datei sofort!</strong>";
    
} catch (Exception $e) {
    echo "✗ Database Error: " . $e->getMessage();
}
?>"""
        
        # Setup-Skript erstellen
        setup_file = LOCAL_ROOT / 'public' / 'SETUP_DATABASE.php'
        with open(setup_file, 'w') as f:
            f.write(setup_script)
        
        # Setup-Skript hochladen
        print("Uploading database setup script...")
        ftp = ftplib.FTP(FTP_CONFIG['host'])
        ftp.login(FTP_CONFIG['user'], FTP_CONFIG['password'])
        upload_file(ftp, str(setup_file), f"{REMOTE_ROOT}/public/SETUP_DATABASE.php")
        ftp.quit()
        
        # Lokale Setup-Datei löschen
        os.remove(setup_file)
        
        print("\n✓ Database setup script uploaded!")
        print("\nWICHTIG: Öffnen Sie jetzt:")
        print("https://firmenpro.de/SETUP_DATABASE.php")
        print("\nund LÖSCHEN Sie die Datei danach sofort!")
        
    except Exception as e:
        print(f"\n✗ Database Error: {str(e)}")
    
    # Phase 3: Abschlussbericht
    print("\n\nPhase 3: Deployment Summary")
    print("-" * 40)
    print("✓ Files uploaded to All-Inkl server")
    print("✓ Directory structure created")
    print("✓ Database setup script ready")
    print("\nNext steps:")
    print("1. Open https://firmenpro.de/SETUP_DATABASE.php")
    print("2. DELETE the setup file immediately after!")
    print("3. Login with admin@firmenpro.de / Admin2025!")
    print("4. Change admin password")
    print("5. Configure cron jobs in All-Inkl KAS")
    print("\nDeployment preparation completed!")

if __name__ == "__main__":
    main()

File: scripts/deploy-to-allinkl.php
#!/usr/bin/env php
<?php

declare(strict_types=1);

/**
 * Deployment-Skript für All-Inkl Hosting
 * 
 * Achtung: Enthält sensible Daten - nach Deployment löschen!
 * 
 * @author 2Brands Media GmbH
 */

echo "===============================================\n";
echo "Metropol Portal - Deployment zu All-Inkl\n";
echo "===============================================\n\n";

// FTP-Konfiguration
$ftpConfig = [
    'server' => 'w019e3c7.kasserver.com',
    'username' => 'w019e3c7',
    'password' => '2Brands2025!',
    'port' => 21,
    'passive' => true,
    'ssl' => false,
    'timeout' => 90,
    'root' => '/w019e3c7/firmenpro.de'
];

// Datenbank-Konfiguration
$dbConfig = [
    'host' => 'localhost',
    'name' => 'd0446399',
    'user' => 'd0446399',
    'password' => '2Brands2025!',
    'charset' => 'utf8mb4'
];

// Projekt-Pfade
$localRoot = dirname(__DIR__);
$remoteRoot = $ftpConfig['root'];

// Verzeichnisse die erstellt werden müssen
$directories = [
    'cache',
    'logs',
    'uploads',
    'temp',
    'backups'
];

// Dateien/Ordner die NICHT hochgeladen werden
$excludePatterns = [
    '.git',
    '.gitignore',
    '.github',
    '.DS_Store',
    'node_modules',
    'tests',
    'docs',
    '*.log',
    '*.md',
    'deploy-to-allinkl.php', // Dieses Skript selbst
    'tsconfig.json',
    'package.json',
    'package-lock.json',
    'composer.lock',
    'phpunit.xml',
    'phpcs.xml',
    'phpstan.neon'
];

/**
 * FTP-Verbindung herstellen
 */
function connectFTP(array $config) {
    echo "Verbinde zu FTP-Server {$config['server']}...\n";
    
    $conn = ftp_connect($config['server'], $config['port'], $config['timeout']);
    if (!$conn) {
        die("Fehler: Konnte keine Verbindung zum FTP-Server herstellen\n");
    }
    
    if (!ftp_login($conn, $config['username'], $config['password'])) {
        die("Fehler: FTP-Login fehlgeschlagen\n");
    }
    
    if ($config['passive']) {
        ftp_pasv($conn, true);
    }
    
    echo "FTP-Verbindung erfolgreich!\n\n";
    return $conn;
}

/**
 * Verzeichnis rekursiv hochladen
 */
function uploadDirectory($conn, $localDir, $remoteDir, $excludePatterns = []) {
    // Remote-Verzeichnis erstellen
    @ftp_mkdir($conn, $remoteDir);
    
    $files = scandir($localDir);
    foreach ($files as $file) {
        if ($file === '.' || $file === '..') {
            continue;
        }
        
        // Prüfen ob Datei/Ordner ausgeschlossen werden soll
        $skip = false;
        foreach ($excludePatterns as $pattern) {
            if (fnmatch($pattern, $file)) {
                $skip = true;
                break;
            }
        }
        if ($skip) {
            continue;
        }
        
        $localPath = $localDir . '/' . $file;
        $remotePath = $remoteDir . '/' . $file;
        
        if (is_dir($localPath)) {
            echo "Erstelle Verzeichnis: $remotePath\n";
            uploadDirectory($conn, $localPath, $remotePath, $excludePatterns);
        } else {
            echo "Upload: $file -> $remotePath\n";
            if (!ftp_put($conn, $remotePath, $localPath, FTP_BINARY)) {
                echo "WARNUNG: Konnte $file nicht hochladen\n";
            }
        }
    }
}

/**
 * Datenbank-Setup
 */
function setupDatabase(array $config) {
    echo "\n===============================================\n";
    echo "Datenbank-Setup\n";
    echo "===============================================\n\n";
    
    try {
        $dsn = "mysql:host={$config['host']};charset={$config['charset']}";
        $pdo = new PDO($dsn, $config['user'], $config['password'], [
            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
            PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC
        ]);
        
        // Datenbank verwenden
        $pdo->exec("USE `{$config['name']}`");
        echo "Datenbank {$config['name']} ausgewählt\n";
        
        // Migrations-Dateien ausführen
        $migrationsPath = dirname(__DIR__) . '/database/migrations';
        $migrations = glob($migrationsPath . '/*.php');
        sort($migrations);
        
        echo "\nFühre Migrationen aus...\n";
        foreach ($migrations as $migration) {
            $name = basename($migration);
            echo "Migration: $name\n";
            
            // Migration ausführen
            require $migration;
            $className = 'Create' . str_replace(['_', '.php'], ['', ''], ucwords(substr($name, 4), '_'));
            if (class_exists($className)) {
                $instance = new $className($pdo);
                if (method_exists($instance, 'up')) {
                    $instance->up();
                }
            }
        }
        
        // Admin-Benutzer anlegen
        echo "\nErstelle Admin-Benutzer...\n";
        $adminPassword = password_hash('Admin2025!', PASSWORD_BCRYPT, ['cost' => 12]);
        $pdo->exec("
            INSERT IGNORE INTO users (name, email, password, role, is_active, created_at, updated_at)
            VALUES ('Administrator', 'admin@firmenpro.de', '$adminPassword', 'admin', 1, NOW(), NOW())
        ");
        
        echo "\nDatenbank-Setup erfolgreich!\n";
        echo "Admin-Login: admin@firmenpro.de / Admin2025!\n";
        echo "WICHTIG: Bitte ändern Sie das Passwort nach dem ersten Login!\n";
        
    } catch (Exception $e) {
        die("Datenbank-Fehler: " . $e->getMessage() . "\n");
    }
}

/**
 * Hauptprogramm
 */
try {
    // FTP-Verbindung
    $ftp = connectFTP($ftpConfig);
    
    // Basis-Verzeichnisse erstellen
    echo "Erstelle Basis-Verzeichnisse...\n";
    foreach ($directories as $dir) {
        $remotePath = $remoteRoot . '/' . $dir;
        @ftp_mkdir($ftp, $remotePath);
        echo "Verzeichnis erstellt: $remotePath\n";
    }
    echo "\n";
    
    // Projekt-Dateien hochladen
    echo "Lade Projekt-Dateien hoch...\n";
    echo "===============================================\n";
    
    // Einzelne Verzeichnisse hochladen
    $uploadDirs = ['src', 'public', 'config', 'database', 'templates', 'lang', 'routes', 'scripts'];
    foreach ($uploadDirs as $dir) {
        if (is_dir($localRoot . '/' . $dir)) {
            echo "\nUpload Verzeichnis: $dir\n";
            uploadDirectory($ftp, $localRoot . '/' . $dir, $remoteRoot . '/' . $dir, $excludePatterns);
        }
    }
    
    // Root-Dateien hochladen
    $rootFiles = ['composer.json'];
    foreach ($rootFiles as $file) {
        if (file_exists($localRoot . '/' . $file)) {
            echo "\nUpload Root-Datei: $file\n";
            ftp_put($ftp, $remoteRoot . '/' . $file, $localRoot . '/' . $file, FTP_BINARY);
        }
    }
    
    // Datei-Berechtigungen setzen
    echo "\nSetze Datei-Berechtigungen...\n";
    @ftp_chmod($ftp, 0755, $remoteRoot . '/cache');
    @ftp_chmod($ftp, 0755, $remoteRoot . '/logs');
    @ftp_chmod($ftp, 0755, $remoteRoot . '/uploads');
    @ftp_chmod($ftp, 0755, $remoteRoot . '/temp');
    
    // FTP-Verbindung schließen
    ftp_close($ftp);
    echo "\nFTP-Upload abgeschlossen!\n";
    
    // Datenbank einrichten
    setupDatabase($dbConfig);
    
    echo "\n===============================================\n";
    echo "Deployment erfolgreich abgeschlossen!\n";
    echo "===============================================\n\n";
    echo "Nächste Schritte:\n";
    echo "1. Öffnen Sie https://firmenpro.de\n";
    echo "2. Loggen Sie sich mit admin@firmenpro.de / Admin2025! ein\n";
    echo "3. Ändern Sie das Admin-Passwort\n";
    echo "4. Konfigurieren Sie E-Mail-Einstellungen\n";
    echo "5. Richten Sie Cron-Jobs ein\n";
    echo "6. LÖSCHEN SIE DIESES DEPLOYMENT-SKRIPT!\n\n";
    
} catch (Exception $e) {
    die("\nFEHLER: " . $e->getMessage() . "\n");
}